lib/Core/Site/QueryType/Base.php

```diff
@@ -0,0 +1,428 @@
+<?php
+
+namespace Netgen\EzPlatformSiteApi\Core\Site\QueryType;
+
+use Closure;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion\LogicalAnd;
+use eZ\Publish\API\Repository\Values\Content\Query\SortClause;
+use InvalidArgumentException;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Base implementation for QueryTypes.
+ *
+ * @internal Do not extend this class directly, extend abstract Content and Location
+ * query types instead.
+ *
+ * @see \Netgen\EzPlatformSiteApi\Core\Site\QueryType\Content
+ * @see \Netgen\EzPlatformSiteApi\Core\Site\QueryType\Location
+ */
+abstract class Base implements QueryType
+{
+    /**
+     * @var \Symfony\Component\OptionsResolver\OptionsResolver
+     */
+    private $optionsResolver;
+
+    /**
+     * @var \Netgen\EzPlatformSiteApi\Core\Site\QueryType\CriterionDefinitionResolver
+     */
+    private $criterionDefinitionResolver;
+
+    /**
+     * @var \Netgen\EzPlatformSiteApi\Core\Site\QueryType\CriteriaBuilder
+     */
+    private $criteriaBuilder;
+
+    /**
+     * @var \Netgen\EzPlatformSiteApi\Core\Site\QueryType\SortClauseParser
+     */
+    private $sortClauseParser;
+
+    /**
+     * @var \Closure[]
+     */
+    private $registeredCriterionBuilders;
+
+    /**
+     * Configure options with the given options $resolver.
+     *
+     * Override this method as needed.
+     *
+     * @param \Symfony\Component\OptionsResolver\OptionsResolver $resolver
+     */
+    protected function configureOptions(OptionsResolver $resolver)
+    {
+        // do nothing
+    }
+
+    /**
+     * Return filter criteria.
+     *
+     * Here you can return null, a single criterion or an array of criteria.
+     * If an array of criteria is returned, it will be combined with base criteria
+     * using logical AND.
+     * Override this method as needed.
+     *
+     * @param array $parameters
+     *
+     * @return null|Criterion|Criterion[]
+     */
+    protected function getFilterCriteria(array $parameters)
+    {
+        return null;
+    }
+
+    /**
+     * Return query criteria.
+     *
+     * Here you can return null or a Criterion instance.
+     * Override this method as needed.
+     *
+     * @param array $parameters
+     *
+     * @return null|Criterion
+     */
+    protected function getQueryCriteria(array $parameters)
```

Comment:
Why is the method called getQueryCriteri**a** (plural) if it returns only single Criterion or null?

Reply:
Right, updated in https://github.com/netgen/ezplatform-site-api/pull/70/commits/841871b4bff40fdfd0298b8b1a3ccefc51d2b905